03_analysis/reader.py

```diff
@@ -0,0 +1,292 @@
+#!/usr/bin/env python
+
+"""
+reader.py
+
+Functions to parse input files or OEMols from those files.
+
+By:      Victoria T. Lim
+Version: Dec 11 2019
+
+"""
+
+import os
+import copy
+import collections
+
+import openeye.oechem as oechem
+from rdkit import Chem, Geometry
+
+def read_mols(infile, mol_slice=None):
+    """
+    Open a molecule file and return molecules and conformers as OEMols.
+    Provide option to slice the mols to return only a chunk from the
+    specified indices.
+
+    Parameters
+    ----------
+    infile : string
+        name of input file with molecules
+    mol_slice : list[int]
+        list of indices from which to slice mols generator for read_mols
+        [start, stop, step]
+
+    Returns
+    -------
+    mols : OEMols
+
+    """
+    ifs = oechem.oemolistream()
+    ifs.SetConfTest(oechem.OEAbsCanonicalConfTest())
+    if not ifs.open(infile):
+        raise FileNotFoundError(f"Unable to open {infile} for reading")
+    mols = ifs.GetOEMols()
+
+    if mol_slice is not None:
+        if len(mol_slice) != 3 or mol_slice[0] >= mol_slice[1] or mol_slice[2] <= 0:
+            raise ValueError("Check input to mol_slice. Should have len 3, "
+                "start value < stop value, step >= 1.")
+
+        # TODO more efficient. can't itertools bc lost mol info (name, SD) after next()
+        # adding copy/deepcopy doesnt work on generator objects
+        # also doesn't work to convert generator to list then slice list
+        #mols = itertools.islice(mols, mol_slice[0], mol_slice[1], mol_slice[2])
+        #mlist = mlist[mol_slice[0]:mol_slice[1]:mol_slice[2]]
+
+        def incrementer(count, mols, step):
+            if step == 1:
+                count += 1
+                return count
+            # use step-1 because for loop already increments once
+            for j in range(step-1):
+                count += 1
+                next(mols)
```

Comment:
I think there might be a bug here. If step==1, then count will increase by 1. If step==2, then count will increase by 1, and the molecule iterator will increment by one.

Reply:
I rewrote the slicing portion to remove this confusion and also to be more robust in terms of what can be used as input.